src/charm.py

```diff
@@ -8,18 +8,223 @@
+            },
+            "serve": {
+                "admin": {
+                    "host": "localhost",
+                    "port": 4445,
+                },
+                "public": {
+                    "host": "localhost",
+                    "port": 4444,
+                },
+            },
+            "urls": {
+                "consent": "http://localhost:3000/consent",
+                "login": "http://localhost:3000/login",
+                "self": {
+                    "issuer": "http://localhost:4444/",
+                    "public": "http://localhost:4444/",
+                },
+            },
+        }
+
+        return yaml.dump(config)
+
+    def _check_leader(self) -> None:
+        if not self.unit.is_leader():
+            # We can't do anything useful when not the leader, so do nothing.
+            raise CheckFailedError("Waiting for leadership", WaitingStatus)
+
+    def _get_database_relation_info(self) -> dict:
+        relation_id = self.database.relations[0].id
+        relation_data = self.database.fetch_relation_data()[relation_id]
+
+        return {
+            "username": relation_data["username"],
+            "password": relation_data["password"],
+            "endpoints": relation_data["endpoints"],
+        }
+
+    def _update_layer(self) -> None:
+        """Updates the Pebble configuration layer and config if changed."""
+        # Get current layer
+        current_layer = self._container.get_plan()
+        # Create a new config layer
+        new_layer = self._hydra_layer
+
+        if current_layer.services != new_layer.services:
+            self.unit.status = MaintenanceStatus("Applying new pebble layer")
+            self._container.add_layer(self._container_name, new_layer, combine=True)
+            logger.info("Pebble plan updated with new configuration")
+            # Push the config on first layer update to avoid PathError
+            self._container.push(self._hydra_config_path, self._config, make_dirs=True)
+            logger.info("Pushed hydra config")
+
+        try:
+            current_config = self._container.pull(self._hydra_config_path).read()
+        except (ProtocolError, PathError) as err:
+            logger.error(str(err))
+            self.unit.status = BlockedStatus(str(err))
+        else:
+            if current_config != self._config:
+                self._container.push(self._hydra_config_path, self._config, make_dirs=True)
+                logger.info("Updated hydra config")
+
+        try:
+            logger.info("Replanning pebble container")
+            self._container.replan()
+        except ChangeError as err:
+            logger.error(str(err))
+            self.unit.status = BlockedStatus("Failed to replan")
+            return
+
+    def _update_container(self, event) -> None:
+        """Update configs, pebble layer and run database migration."""
+        try:
+            self._check_leader()
+        except CheckFailedError as err:
```

Comment:
I don't think checking leadership needs to be in a helper function. It's only used here. We can handle it like all the other checks.